File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    // find size
    let size = 0;
    let curr = head;
    
    while(curr != null) {
        size++;
        curr = curr.next;
    }
    
    if(size === 1) return null;
    
    // intead of getting the n node from last we can update n from front
    n = size - n + 1;
    
    // if first element needs to be removed
    if(n === 1) return head.next;
    
    curr = head;
    let prev = null;
    let count = 0;
    
    while(curr != null) {
        count++;
        if(count === n) {
            prev.next = curr.next;
        } else {
            prev = curr;
        }
        
        curr = curr.next;
    }
    
    return head;
};

// tc -> O(n)
// sc -> O(1)

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
    // remove nth from end means remove len - n + 1 from front
    cNode := head
    count := 0
    var temp *ListNode
    
    for cNode != nil {
        count++
        cNode = cNode.Next
    }
    
    // set the new value m which will be just one before the node to be removed
    m := count - n
    cNode = head
    
    // if m is zero means the first head needs to be removed
    if m == 0 {
        temp = cNode.Next
        cNode.Next = nil
        return temp
    }
    
    // get to one node before the one that is to be removed
    for m > 1 {
        cNode = cNode.Next
        m--;
    }
    
    // remove operation
    temp = cNode.Next
    cNode.Next = temp.Next
    temp.Next = nil
    
    return head
}